sys/kern/mcount.c

```diff
@@ -0,0 +1,195 @@
+ *             (0 - means the end of the list)
+ * 
+ *  tos[0] + 1  is the smallest index of an unused tos entry
+ */
+
+void __cyg_profile_func_enter(void *from, void *self) {
+  u_long frompc = (u_long)from, selfpc = (u_long)self;
+  u_short *frompcindex;
+  tostruct_t *top, *prevtop;
+  gmonparam_t *p = &_gmonparam;
+  long toindex;
+
+  /*
+   * check that we are profiling
+   * and that we aren't recursively invoked.
+   */
+  if (p->state != GMON_PROF_ON)
+    return;
+
+  WITH_SPIN_LOCK (&mcount_lock) {
+    p->state = GMON_PROF_BUSY;
+    /*
+     * check that frompc is a reasonable pc value.
+     * for example:	signal catchers get called from the stack,
+     *		not from text space.  too bad.
+     */
+    frompc -= p->lowpc;
+    if (frompc >= p->textsize)
+      goto done;
+
+    size_t index = (frompc / (HASHFRACTION * sizeof(*p->froms)));
+    frompcindex = &p->froms[index];
+    toindex = *frompcindex;
+    /*
+     *	First time profiling this calling function .
+     */
+    if (toindex == 0) {
+      /*
+       * Getting an unused node (the smallest unused tos entry index).
+       */
+      toindex = ++p->tos[0].link;
+      if (toindex >= p->tolimit) {
+        p->state = GMON_PROF_ERROR;
+        goto done;
+      }
+
+      *frompcindex = (u_short)toindex;
+      top = &p->tos[(size_t)toindex];
+      top->selfpc = selfpc;
+      top->count = 1;
+      top->link = 0;
+      goto done;
+    }
+    top = &p->tos[(size_t)toindex];
+    /*
+     * Node with our called function at front of chain; usual case.
+     */
+    if (top->selfpc == selfpc) {
+      top->count++;
+      goto done;
+    }
+    /*
+     * Traversing the list and looking for node with our called function.
+     */
+    while (true) {
+      /*
+       * We reached the end of the list it does not contain a node with
+       * the called function. Check if there are still available nodes to use,
+       *  if so get one and add it to the list.
+       */
+      if (top->link == 0) {
+        toindex = ++p->tos[0].link;
+        if (toindex >= p->tolimit) {
+          p->state = GMON_PROF_ERROR;
+          goto done;
+        }
+
+        top = &p->tos[(size_t)toindex];
+        top->selfpc = selfpc;
+        top->count = 1;
+        top->link = *frompcindex;
+        *frompcindex = (u_short)toindex;
+        goto done;
+      }
+      /*
+       * Move to the next node.
+       */
+      prevtop = top;
+      top = &p->tos[top->link];
+      /*
+       * We found our node,remove it from our list
+       * and add it at the beginning of the list.
+       */
+      if (top->selfpc == selfpc) {
+        top->count++;
+        toindex = prevtop->link;
+        prevtop->link = top->link;
+        top->link = *frompcindex;
+        *frompcindex = (u_short)toindex;
+        goto done;
+      }
+    }
+  done:
+    if (p->state != GMON_PROF_ERROR)
+      p->state = GMON_PROF_ON;
+  }
+}
+
+void __cyg_profile_func_exit(void *this_fn, void *call_site) {
+}
```

Comment:
Is it intentional? Do you have any future plans to use that function?

Reply:
It is intentional, we need to provide this function to use  `-finstrument-functions`